service/scaladsl/api/src/main/scala/com/lightbend/lagom/scaladsl/api/deser/PathParamSerializer.scala

```diff
@@ -99,4 +101,25 @@ object PathParamSerializer {
       }
     }
   }
+
+  /**
+   * A traversable path param serializer
+   */
+  implicit def traversablePathParamSerializer[CC[X] <: Traversable[X], Param: PathParamSerializer](implicit bf: CanBuildFrom[CC[_], Param, CC[Param]]): PathParamSerializer[CC[Param]] = {
+    val delegate = implicitly[PathParamSerializer[Param]]
+    val name = delegate match {
+      case named: NamedPathParamSerializer[Param] => s"Traversable[${named.name}]"
+      case other                                  => s"Traversable($other)"
+    }
+
+    new NamedPathParamSerializer[CC[Param]](name) {
+      override def serialize(parameter: CC[Param]): Seq[String] = parameter.flatMap(delegate.serialize).to[Seq]
+      override def deserialize(parameters: Seq[String]): CC[Param] = {
+        val builder = bf()
+        builder.sizeHint(parameters)
+        builder ++= parameters.map(param => delegate.deserialize(Seq(param)))
+        builder.result()
+      }
+    }
+  }
```

Comment:
Move this to a new trait called `LowPriorityPathParamSerializers`, and make `PathParamSerializer` extend it. This can have a significant improvement on compiler performance.

See https://github.com/playframework/playframework/pull/4313 for more details.

Reply:
Agree. Preferences on what to do and where to place the `NamedPathParamSerialized` (currently) private class? Lift it outside, make it public for anyone to use, `sealed abstract` within the new `LowPriorityPathParamSerializers` trait, other?

Reply:
@sirocchj having it defined on that same file is fine. The goal is to have the object `PathParamSerializer` to extend it as in the play example pointed out by @jroper, so it needs to leave outside it. 

I won't make it public in any case. This new trait is not meant for extension, but for efficient implicit lookup. 

Btw, the `optionPathParamSerializer` can also move to `LowPriorityPathParamSerializers`.

Thanks

Reply:
@sirocchj I did some research and tests after our discussion on Gitter. I compiled with `-Xlog-implicits` in order to be able to inspect the logs.

Here is what I found:

`optionPathParamSerializer` should not be at same level as `traversablePathParamSerializer`. So not inside `LowPriorityPathParamSerializers` as I suggested on my previous message. 
When put together, we get a similar logging as the one shown in https://github.com/playframework/playframework/pull/4313

Having the base implicits on a `DefaultPathParamSerializers` doesn't make any difference. It's more a matter of taste / style. I personally prefer to have the defaults on their own trait like you suggested. 

### And there is somthing else...

Because we now have two implicits for `PathParamSerializer[F[Param]]`, ie: `PathParamSerializer[Option[Param]]` and `PathParamSerializer[CC[Param]]`,  and because we are using Context Bounds, we get two extra implicit lookups resulting in two `hasMatchingSymbol` errors.

This happens because of the combination of Context Bound and `implicitly` lookup. 

We should instead remove the Context Bound and add a explicit implicit arg, for instance:
```scala
implicit def optionPathParamSerializer[Param]
(implicit delegate: PathParamSerializer[Param]) = ...

implicit def traversablePathParamSerializer[CC[X] <: Traversable[X], Param]
(implicit bf: CanBuildFrom[CC[_], Param, CC[Param]],  delegate: PathParamSerializer[Param]) = ...
```

With respect to Context Bound my rule of thumb is to avoid it when I need to use the implicit in the method body itself. In other words, if the reason to have an implicit is to pass it further (implicitly) to another method, and without referring to it explicitly, than  a Context Bound is perfectly fine.

Reply:
@sirocchj if you want to play around with the available options, [here is a small sample](https://github.com/lagom/lagom/files/1359313/scala-implicits.zip) I used to run the tests I just mentioned.